### PR TITLE
Embolden

### PR DESCRIPTION
**Pull request by behdad**
Fixes https://github.com/harfbuzz/harfbuzz/issues/3883

### CODE REVIEW

**Comment by behdad** [comment]
Todo:
- [ ] Relicense the code from FreeType authors (contacted),
- [ ] Add `hb_font_t` API and hook it up. Currently embolden amount hardcoded.
- [ ] Adapt advances and shaping to it.

**Comment by matthiasclasen** [comment]
I've added some docs.

**Comment by behdad** [comment]
@khaledhosny rightly pointed out that adjusting the advance-widths unfortunately breaks Arabic fonts that use rounded edges for their connecting strokes. We can't ignore that unfortunately. So I'm stuck as to how to proceed. Khaled also mentioned that in LibreOffice he uses unadjusted advance-widths and so far no one complained.

If using unadjusted advance-widths though, I think we should NOT shift the glyph shape horizontally then. Which I can do. But then we should also unshift the FreeType outlines in hb-ft.  I think we should still shift them vertically such that they sit on the baseline.

All of this ends up who knows where in vertical writing....

**Comment by behdad** [comment]
No idea why on mac-os two shape tests timeout. Is it just that bot is so underpowered?!

**Comment by khaledhosny** [comment]
> @khaledhosny rightly pointed out that adjusting the advance-widths unfortunately breaks Arabic fonts that use rounded edges for their connecting strokes. We can't ignore that unfortunately. So I'm stuck as to how to proceed. Khaled also mentioned that in LibreOffice he uses unadjusted advance-widths and so far no one complained.
> 
> If using unadjusted advance-widths though, I think we should NOT shift the glyph shape horizontally then. Which I can do. But then we should also unshift the FreeType outlines in hb-ft. I think we should still shift them vertically such that they sit on the baseline.
> 
> All of this ends up who knows where in vertical writing....

There are other cases where keeping advance width is desired like monospace fonts or using emboldening to achieve something like [grade](https://fonts.google.com/knowledge/glossary/grade). I think it might be worth while to add an option or a flag to control advance adjustment and shift in both directions.

**Comment by behdad** [comment]
> There are other cases where keeping advance width is desired like monospace fonts or using emboldening to achieve something like [grade](https://fonts.google.com/knowledge/glossary/grade). I think it might be worth while to add an option or a flag to control advance adjustment and shift in both directions.

Matthias had the same idea. Okay I'm convinced.

**Comment by behdad** [comment]
Okay so then to hb-view I'll add `--font-bold` and `--font-grade`?

**Comment by behdad** [comment]
@khaledhosny @matthiasclasen How does this look?

**Comment by behdad** [comment]
> * [ ]  Glyph extents.
> * [ ]  Font extents. The rest of metrics we're not going to try.

Working on these.

**Comment by behdad** [comment]
Should grade shift glyphs up or not? Currently it does. @khaledhosny @davelab6

**Comment by behdad** [comment]
> > * [ ]   Glyph extents.
> > * [ ]   Font extents. The rest of metrics we're not going to try.
> 
> Working on these.

Done. I think this is good to land now. I can work on `hb-ot-metrics` later.

File: src/hb-font.cc [hunk 1]
```diff
@@ -2442,6 +2451,74 @@ hb_font_get_ptem (hb_font_t *font)
...
+ * 0.01 to 0.05 range. The default value is zero.
+ *
+ * Synthetic boldness is applied by offsetting the contour
+ * points of the glyph shape.
+ *
+ * Synthetic boldness is applied when rendering a glyph via
+ * hb_font_draw_glyph().
+ *
+ * If @in_place is %FALSE, then glyph advance-widths are also
+ * adjusted, otherwise they are not.
```
Comment (matthiasclasen) [inline src/hb-font.cc:2474]: It might be nice to give a hint about the use case for @in_place here.

File: util/font-options.hh [hunk 2]
```diff
@@ -286,7 +332,11 @@ font_options_t::add_options (option_parser_t *parser)
 			      G_OPTION_ARG_CALLBACK,	(gpointer) &parse_font_ppem,	"Set x,y pixels per EM (default: 0; disabled)",	"1/2 integers"},
     {"font-ptem",	0, font_size_flags,
 			      G_OPTION_ARG_DOUBLE,	&this->ptem,			"Set font point-size (default: 0; disabled)",	"point-size"},
-    {"font-slant",	0, 0,
+    {"font-bold",	0, font_size_flags,
+			      G_OPTION_ARG_CALLBACK,	(gpointer) &parse_font_bold,	"Set synthetic bold (default: 0)",		"1/2 numbers; eg. 0.05"},
+    {"font-grade",	0, font_size_flags,
```
Comment (matthiasclasen) [inline util/font-options.hh:337]: Is that a well-known term? I would not know what --font-grade does.

**Comment by matthiasclasen** [comment]
Looks good to me apart from those two comments.

File: util/font-options.hh (hunk 2 above)
Comment (behdad) [reply util/font-options.hh:337]: It is. I just don't know where to explain it more.

File: src/hb-font.cc (hunk 1 above)
Comment (behdad) [reply src/hb-font.cc:2474]: I'll mention grading here and link to https://fonts.google.com/knowledge/glossary/grade

Comment (behdad) [reply src/hb-font.cc:2474]: @khaledhosny What's the docbook syntax for links? https://developer-old.gnome.org/gtk-doc-manual/stable/documenting_syntax.html.en only has the markdown syntax. Is that kosher in our docs?

Comment (behdad) [reply src/hb-font.cc:2474]: > Is that kosher in our docs?

Seems to work.

**Pull request merged**